core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala

```diff
@@ -236,13 +265,25 @@ class HadoopMapReduceCommitProtocol(
     committer.setupTask(taskContext)
     addedAbsPathFiles = mutable.Map[String, String]()
     partitionPaths = mutable.Set[String]()
+    dynamicStagingTaskFiles = mutable.Set[Path]()
   }
 
   override def commitTask(taskContext: TaskAttemptContext): TaskCommitMessage = {
     val attemptId = taskContext.getTaskAttemptID
     logTrace(s"Commit task ${attemptId}")
     SparkHadoopMapRedUtil.commitTask(
       committer, taskContext, attemptId.getJobID.getId, attemptId.getTaskID.getId)
+    if (dynamicPartitionOverwrite) {
+      val fs = stagingDir.getFileSystem(taskContext.getConfiguration)
+      dynamicStagingTaskFiles.foreach { stagingTaskFile =>
+        val fileName = stagingTaskFile.getName
+        val taskPartitionPath = getDynamicPartitionPath(stagingTaskFile, taskContext)
+        val destFile = new Path(taskPartitionPath, fileName)
+        if (!fs.rename(stagingTaskFile, destFile)) {
```

Comment:
we need to check `destFile.exist()` before `rename()`?

Reply:
I think we can add a check here, and if destFile existed, I think we can remove it directly, for that this operation is after SparkHadoopMapRedUtil.commitTask.
If outputCommitCoordination is enabled, only one task can commit its output.
Otherwise, it does not matter to remove it directly.

Reply:
No, I mean, according to the basic idea, we should skip rename if a previous task attempt has already renamed once(so `destFile` should be existed), no?

Reply:
got, I will add a check. thanks